runtime/weak.c

```diff
@@ -537,10 +574,33 @@ CAMLexport void caml_ephemeron_blit_key(value ars, mlsize_t offset_s,
   offset_s += CAML_EPHE_FIRST_KEY;
   offset_d += CAML_EPHE_FIRST_KEY;
 
+  if ( caml_gc_phase == Phase_mark
+       && caml_ephe_list_pure
+       && Field(ard, CAML_EPHE_DATA_OFFSET) != caml_ephe_none
+       && !Is_white_val(ard)
+       ){
+    /* If the destination is in set (2) it is because of a white key */
+
+    dest_has_white_value = 0;
+
+    for(i = 0; i < length; i++){
+      dest_has_white_value |= Is_White_During_Mark(Field(ard, offset_d + i));
+    };
+    /* test if the destination can't be in set (2) because of the keys that are going to be set */
+    if(!dest_has_white_value) goto No_darkening;
+    for(i = 0; i < length; i++){
+      /* test if the source is going to bring a white key to replace the one set */
+      if(Is_White_During_Mark(Field(ars, offset_s + i))) goto No_darkening;
+    };
+    /* the destination ephemeron could be in the set (2) because of a white key replaced and not have one anymore after. */
```

Comment:
```suggestion
    /* the destination ephemeron could be in the set (2) because of a white key
        replaced and not have one anymore after. */
```